In [1]:
import numpy as np
import pandas as pd

In [2]:
hawks_df = pd.read_csv('./hawks.csv')
hawks_df.head()

Month  Day  Year CaptureTime ReleaseTime  BandNumber Species Age  Sex  \
0      9   19  1992       13:30         NaN   877-76317      RT   I  NaN   
1      9   22  1992       10:30               877-76318      RT   I  NaN   
2      9   23  1992       12:45               877-76319      RT   I  NaN   
3      9   23  1992       10:50               745-49508      CH   I    F   
4      9   27  1992       11:15              1253-98801      SS   I    F   

    Wing  Weight  Culmen  Hallux  Tail  StandardTail  Tarsus  WingPitFat  \
0  385.0   920.0    25.7    30.1   219           NaN     NaN         NaN   
1  376.0   930.0     NaN     NaN   221           NaN     NaN         NaN   
2  381.0   990.0    26.7    31.3   235           NaN     NaN         NaN   
3  265.0   470.0    18.7    23.5   220           NaN     NaN         NaN   
4  205.0   170.0    12.5    14.3   157           NaN     NaN         NaN   

   KeelFat  Crop  
0      NaN   NaN  
1      NaN   NaN  
2      NaN   NaN  
3      NaN   NaN  
4      NaN   NaN

In [3]:
shape = hawks_df.shape 
shape

(908, 19)

In [4]:
lst_missing_ratio = []

for i in range(0,shape[1]):    
    column = hawks_df[hawks_df.columns[i]]
    miss_1 = (column == ' ').sum()
    miss_2 = (column.isna()).sum()
    miss_ratio = ((miss_1 + miss_2) / shape[0] * 100).round(1)
    print(miss_ratio,hawks_df.columns[i])    
    lst_missing_ratio.append(miss_ratio)


0.0 Month
0.0 Day
0.0 Year
0.1 CaptureTime
92.8 ReleaseTime
0.2 BandNumber
0.0 Species
0.0 Age
63.4 Sex
0.1 Wing
1.1 Weight
0.8 Culmen
0.7 Hallux
0.0 Tail
37.1 StandardTail
91.7 Tarsus
91.5 WingPitFat
37.6 KeelFat
37.8 Crop


C:\Users\phamd\anaconda3\lib\site-packages\pandas\core\ops\array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


- Đối với những thuộc tính bị thiếu trên 50% thì xoá luôn thuộc tính đi  
    - ReleaseTime: thiếu 92.8%
    - Sex: thiếu 63.4%
    - Tarsus: thiếu 91.7%
    - WingPitFat: thiếu 91.5%
    
- Các giá trị thiếu của thuộc tính numeric sẽ được điền đầy bằng phương pháp mean
    - CaptureTime
    - Wing 
    - Weight
    - Culmen 
    - Hallux 
    - StandardTail
    - KeelFat 
    - Crop
- Các giá trị thiếu của thuộc tính nominal được điền đầy bằng phương pháp mode:
    - CaptureTime 
    
- Ta thấy thuộc tính BandNumber là thuộc tính index vì tất cả các dòng dữ liệu đều khác nhau. Do đó, ta xoá luôn cột này

In [5]:
# xoá cột
hawks_df.drop(columns=['ReleaseTime', 'Sex', 'Tarsus', 'WingPitFat', 'BandNumber'], inplace=True)
hawks_df.head()

Month  Day  Year CaptureTime Species Age   Wing  Weight  Culmen  Hallux  \
0      9   19  1992       13:30      RT   I  385.0   920.0    25.7    30.1   
1      9   22  1992       10:30      RT   I  376.0   930.0     NaN     NaN   
2      9   23  1992       12:45      RT   I  381.0   990.0    26.7    31.3   
3      9   23  1992       10:50      CH   I  265.0   470.0    18.7    23.5   
4      9   27  1992       11:15      SS   I  205.0   170.0    12.5    14.3   

   Tail  StandardTail  KeelFat  Crop  
0   219           NaN      NaN   NaN  
1   221           NaN      NaN   NaN  
2   235           NaN      NaN   NaN  
3   220           NaN      NaN   NaN  
4   157           NaN      NaN   NaN

In [6]:
# Đối với các cột numeric: thay missing values bằng mean

num_col = ['Wing' ,'Weight','Culmen' ,'Hallux' ,'StandardTail','KeelFat' ,'Crop']

for item in num_col:
    hawks_df[item] = hawks_df[item].replace(' ', np.nan, regex=True)
    hawks_df[item] = hawks_df[item].fillna(hawks_df[item].mean())

hawks_df.head()

Month  Day  Year CaptureTime Species Age   Wing  Weight     Culmen  \
0      9   19  1992       13:30      RT   I  385.0   920.0  25.700000   
1      9   22  1992       10:30      RT   I  376.0   930.0  21.801498   
2      9   23  1992       12:45      RT   I  381.0   990.0  26.700000   
3      9   23  1992       10:50      CH   I  265.0   470.0  18.700000   
4      9   27  1992       11:15      SS   I  205.0   170.0  12.500000   

      Hallux  Tail  StandardTail   KeelFat     Crop  
0  30.100000   219    199.182137  2.184303  0.23454  
1  26.410865   221    199.182137  2.184303  0.23454  
2  31.300000   235    199.182137  2.184303  0.23454  
3  23.500000   220    199.182137  2.184303  0.23454  
4  14.300000   157    199.182137  2.184303  0.23454

In [7]:
# Đối với các cột norminal: thay missing values bằng mode
hawks_df['CaptureTime'] = hawks_df['CaptureTime'].replace(' ', np.nan, regex=True)
hawks_df['CaptureTime'] = hawks_df['CaptureTime'].fillna((hawks_df['CaptureTime'].mode()[0]))
hawks_df.head()


Month  Day  Year CaptureTime Species Age   Wing  Weight     Culmen  \
0      9   19  1992       13:30      RT   I  385.0   920.0  25.700000   
1      9   22  1992       10:30      RT   I  376.0   930.0  21.801498   
2      9   23  1992       12:45      RT   I  381.0   990.0  26.700000   
3      9   23  1992       10:50      CH   I  265.0   470.0  18.700000   
4      9   27  1992       11:15      SS   I  205.0   170.0  12.500000   

      Hallux  Tail  StandardTail   KeelFat     Crop  
0  30.100000   219    199.182137  2.184303  0.23454  
1  26.410865   221    199.182137  2.184303  0.23454  
2  31.300000   235    199.182137  2.184303  0.23454  
3  23.500000   220    199.182137  2.184303  0.23454  
4  14.300000   157    199.182137  2.184303  0.23454

In [8]:
# lưu file

# set nominal attribute as class attribute
hawks_df = hawks_df[['Month', 'Day', 'Year', 'CaptureTime', 'Age', 'Wing','Weight',
                     'Culmen', 'Hallux', 'Tail', 'StandardTail', 'KeelFat', 'Crop', 'Species']] 

hawks_df.to_csv('Preprocess.csv', header=True, index=False)